lib/remote_node.go

```diff
@@ -0,0 +1,468 @@
+	// the height of the latest actual block you have. This makes it so that
+	// peers who have up-to-date headers but missing blocks won't be considered
+	// for initial block download.
+	ver.StartBlockHeight = uint32(rn.bc.BlockTip().Header.Height)
+
+	// Set the minimum fee rate the peer will accept.
+	ver.MinFeeRateNanosPerKB = rn.minTxFeeRateNanosPerKB
+
+	return ver
+}
+
+// HandleVersionMessage is called upon receiving a version message from the RemoteNode's peer. The peer may be the one
+// initiating the handshake, in which case, we should respond with our own version message. To do this, we pass the
+// responseNonce to this function, which we will use in our response version message.
+func (rn *RemoteNode) HandleVersionMessage(verMsg *MsgDeSoVersion, responseNonce uint64) {
+	if verMsg.Version < rn.params.MinProtocolVersion {
+		glog.V(1).Infof("RemoteNode.HandleVersionMessage: Requesting PeerDisconnect for id: (%v) "+
+			"protocol version too low. Peer version: %v, min version: %v", rn.peer.ID, verMsg.Version, rn.params.MinProtocolVersion)
+		rn.Disconnect()
+		return
+	}
+
+	vMeta := rn.GetHandshakeMetadata()
+	if vMeta.versionTimeExpected != nil && vMeta.versionTimeExpected.Before(time.Now()) {
+		glog.V(1).Infof("RemoteNode.HandleVersionMessage: Requesting PeerDisconnect for id: (%v) "+
+			"version timeout. Time expected: %v, now: %v", rn.peer.ID, vMeta.versionTimeExpected.UnixMicro(), time.Now().UnixMicro())
+		rn.Disconnect()
+		return
+	}
+
+	// Save the version nonce so we can include it in our verack message.
+	vMeta.versionNoncesReceived = verMsg.Nonce
+
+	// Set the peer info-related fields.
+	vMeta.userAgent = verMsg.UserAgent
+	vMeta.ServiceFlag = verMsg.Services
+	vMeta.advertisedProtocolVersion = NewProtocolVersionType(verMsg.Version)
+	negotiatedVersion := rn.params.ProtocolVersion
+	if verMsg.Version < rn.params.ProtocolVersion.ToUint64() {
+		negotiatedVersion = NewProtocolVersionType(verMsg.Version)
+	}
+	vMeta.negotiatedProtocolVersion = negotiatedVersion
+	vMeta.minTxFeeRateNanosPerKB = verMsg.MinFeeRateNanosPerKB
+	timeConnected := time.Unix(verMsg.TstampSecs, 0)
+	vMeta.timeConnected = &timeConnected
+	vMeta.timeOffsetSecs = verMsg.TstampSecs - time.Now().Unix()
+	vMeta.StartingBlockHeight = verMsg.StartBlockHeight
+
+	// Update the timeSource now that we've gotten a version message from the peer.
+	rn.cmgr.AddTimeSample(rn.peer.Address(), timeConnected)
+
+	if !rn.peer.IsOutbound() {
+		// Respond to the version message if this is an inbound peer.
+		rn.sendVersionMessage(responseNonce)
+	}
+	// After sending and receiving a compatible version, complete the
+	// negotiation by sending and receiving a verack message.
+	verackTimeExpected := time.Now().Add(rn.params.VersionNegotiationTimeout)
+	vMeta.verackTimeExpected = &verackTimeExpected
+	if err := rn.sendVerack(); err != nil {
+		glog.Errorf("RemoteNode.HandleVersionMessage: Problem sending verack message to peer (id= %d): %v", rn.peer.ID, err)
```

Comment:
I generally prefer to return errors and require the caller to handle them instead of glogging them. It just becomes tricky to track down bugs later on.